### PR TITLE
Support for packer v1.7.0

### LINKED ISSUE #948: Add support for packer init, introduced in packer 1.7 and required after packer 2.x
Starting with packer 1.7, packer has broken out the plugins from the packer source code. This means that the templates will need to list out the required plugins like terraform (see the [upgrade guide](https://www.packer.io/guides/1.7-template-upgrade)) and run `packer init` to pull down the plugins.

The terratest functions for packer should be updated to work with this new model. Right now, 1.7 is still backward compatible with the old assumption of baked in plugins, but in the next version this will no longer be true.

### PR DESCRIPTION
**Pull request by bwhaley**
Fixes #948.

Adds support for running `packer init` in Packer version >= v1.7.0. A few comments about this change:

- This should be backwards compatible with Packer < v1.7.0.
- All the example Packer template are converted to HCL. I've kept the JSON equivalents in the repo for now since [the old format is still supported](https://www.packer.io/guides/1.7-template-upgrade).
- Introduces a new dependency, [hashicorp/go-version](https://github.com/hashicorp/go-version). Hopefully this is low impact/low overhead.
- It turns out that when downloading plugins, Packer will happily stomp all over itself, causing checksum problems and difficult-to-debug failure modes if multiple tests are running and using the same plugins (e.g. the hashicorp/amazon plugin) at about the same time. To mitigate this, Terratest will instruct Packer to download plugins to a temporary directory, then will clean up when finished. This behavior is the default, but users can set [`DisableTemporaryPluginPath`](https://github.com/gruntwork-io/terratest/blob/85786f320cad3d87231f0aefe318a654e10696e0/modules/packer/packer.go#L39) to disable it if desired.
- The empty S3 bucket tests were unreasonably slow, so I lowered the file count from 10,000 to 1,000. Hopefully that should still be a sufficient quantity to remain an effective test.

All comments & feedback welcome!

### CODE REVIEW

File: modules/packer/packer.go [hunk 1]
```diff
@@ -92,6 +97,30 @@ func BuildArtifact(t testing.TestingT, options *Options) string {
...
+	// to download the same plugin at the same time to the global path.
+	// Set DisableTemporaryPluginPath to disable this behavior.
+	if !options.DisableTemporaryPluginPath {
+		// The built-in env variable defining where plugins are downloaded
+		const packerPluginPathEnvVar = "PACKER_PLUGIN_PATH"
+		options.Logger.Logf(t, "Creating a temporary directory for Packer plugins")
+		pluginDir, err := ioutil.TempDir("", "terratest-packer-")
+		if err != nil {
+			log.Fatal(err)
+		}
```
Comment (yorinasub17) [inline modules/packer/packer.go:111]: NIT: use `require.NoError`

**Review by yorinasub17 — APPROVED** [review]
Had a few nits, but overall LGTM! You can choose to address the nits in this PR, or in a follow up PR.

Feel free to merge as is (assuming the build passes!)

File: modules/packer/packer.go [hunk 2]
```diff
@@ -145,6 +174,73 @@ func extractArtifactID(packerLogOutput string) (string, error) {
...
+	cmd := shell.Command{
+		Command:    "packer",
+		Args:       []string{"-version"},
+		Env:        options.Env,
+		WorkingDir: options.WorkingDir,
+	}
+	description := fmt.Sprintf("%s %v", cmd.Command, cmd.Args)
+	localVersion, err := retry.DoWithRetryableErrorsE(t, description, options.RetryableErrors, options.MaxRetries, options.TimeBetweenRetries, func() (string, error) {
+		return shell.RunCommandAndGetOutputE(t, cmd)
+	})
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply modules/packer/packer.go]: Oh, good point! Fixed in 950fddf5.

File: modules/packer/packer.go [hunk 3]
```diff
@@ -145,6 +174,73 @@ func extractArtifactID(packerLogOutput string) (string, error) {
...
+	options.Logger.Logf(t, "Running Packer init")
+
+	cmd := shell.Command{
+		Command:    "packer",
+		Args:       []string{"init", options.Template},
+		Env:        options.Env,
+		WorkingDir: options.WorkingDir,
+	}
+
+	description := fmt.Sprintf("%s %v", cmd.Command, cmd.Args)
```
[the remark this answers is not in the document]
Comment (bwhaley) [reply modules/packer/packer.go]: Done: 950fddf5

File: modules/packer/packer.go (hunk 1 above)
Comment (bwhaley) [reply modules/packer/packer.go]: Done: 950fddf5

**Comment by bwhaley** [comment]
Thanks for the review. I've incorporated the feedback. Relevant tests are passing, but I notice that `TestGitRefChecks/GetCurrentRefReturnsLightTagValue` is failing with:

```
 git_test.go:56: 
        	Error Trace:	git_test.go:56
        	Error:      	Not equal: 
        	            	expected: "v0.0.1-1-g58d3ea8"
        	            	actual  : "v0.0.1-1-g58d3ea8f"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-v0.0.1-1-g58d3ea8
        	            	+v0.0.1-1-g58d3ea8f
        	Test:       	TestGitRefChecks/GetCurrentRefReturnsLightTagValue
    --- FAIL: TestGitRefChecks/GetCurrentRefReturnsLightTagValue (0.02s)
```

Is this a known testing error?

**Comment by yorinasub17** [comment]
> Is this a known testing error?

I don't think this is a known error given that [it passed recently](https://app.circleci.com/pipelines/github/gruntwork-io/terratest/697/workflows/a5bee147-a3a7-4d97-ac8b-941b5249d304/jobs/6162). I think this might be a new error caused by overlapping git commit SHA prefixes that caused it to require another character to differentiate that one?

**Review by yorinasub17 — APPROVED** [review]
Updates LGTM!

**Comment by bwhaley** [comment]
Thanks! Relevant tests are passing. Will submit a new PR to fix the `GetCurrentRefReturnsLightTagValue` test.

**Review by brikis98 — COMMENTED** [review]
Ben, it looks like a few of the Packer templates were deleted... But these were used on the Terratest home page as examples: https://terratest.gruntwork.io/#test-with-terratest. If you click "Packer" and the "build.json" tab, you get an error. Could you update those examples to use the HCL files instead?

**Comment by bwhaley** [comment]
Doh, I thought I'd found all the references. Opened https://github.com/gruntwork-io/terratest/pull/966 to fix.

**Pull request merged**